design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,319 @@
+resource, if the project does indeed want to migrate.
+
+
+### Goals
+
+* Add `CertificateSigningRequest` signer controllers to cert-manager for all
+    `Issuer` types
+* Pave a way for replacement of the `CertificateRequest` resource for `CertificateSigningRequest`
+
+### Non-Goals
+
+* Remove the `CertificateRequest` resource before `CertificateSigningRequest` is
+    fully supported and a migration path is in place
+
+---
+
+## Reasons for Approach
+
+Below are a summary of properties of the `CertificateSigningRequest` which
+cause issues or caveats which need to be addressed when adding support for the
+resource.
+
+### Read all access to CertificateSigningRequests
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. This means that in order for users to have the
+same debugging and usage capabilities of the `CertificateRequest`, users will
+have the ability to list all requests in the cluster, rather than being scoped to
+specific namespaces. This is a regression in security.
+
+A solution to this problem could be given to users by providing examples or
+guides to external tooling (e.g. [Open Policy Agent
+(OPA)](https://github.com/open-policy-agent/opa), whereby cluster administrators
+can limit users access to certain `CertificateSigningRequests` based on label
+selectors.
+
+
+### Namespaced issuers referenced by CertificateSigningRequest
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. cert-manager has the concept of namespaced
+issuers that may only be referenced by `CertificateRequests` that reside in
+the same namespace. Since `CertificateSigningRequests` can reference any
+namespaced `Issuer`, there must be some mechanism to prevent privilege
+escalation (Alice references an issuer in Bob's namespace).
+
+cert-manager will enforce an
+[RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) noun and
+verb whereby the requester must have this role bound to them, in order for the
+`CertificateSigningRequest` referencing a namespaced `Issuer` be approved by the
+cert-manager controller. See [here](#conditions).
+
+This will be done via a
+[`SubjectAccessReview`](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/authorization/v1/types.go#L52)
+against the `UserInfo` fields on the `CertificateSigningRequest`. The required
+role must be of `certificates.k8s.io/CertificateSigningRequest` with the verb
+`create` within the target namespace.
+
+Since `Certificates` are also namespaced resources, cert-manager validates that
+created `Certificate` resources may only reference `Issuers` in the same
+namespace, or `ClusterIssuers`. Any resulting `CertificateSigningRequests` will
+reference `Issuers` in the same namespace.
+
+### Duration Field
+
+`CertificateSigningRequests` do not include a `duration` field. To have parity
```

Comment:
I am working on adding a `NotAfterHint` field to the Kube CSR resource.

Reply:
https://github.com/kubernetes/kubernetes/pull/99494

Reply:
This has some implications on how conversion would work (time.Duration <-> time.Time). I _think_ it should be possible if we have both the `Duration` and `NotAfterHint` present on the resource, but is not trivial.

Reply:
I think it would be straightforward if you just use the creation timestamp on the Kube CSR object?  If you want to keep using the existing duration based approach, then the client would have to be a bit smarter about how it makes the CSR request.  I do not know how exact you all want to make this transition - depending on wherever Kube lands on this, would it be easier to just switch to the Kube approach (i.e. specify the end time instead of the duration)?

Reply:
Circling back on this, the upstream proposal has shifted to use a duration instead of `NotAfterHint`, which I think means this can be resolved? 😄 https://github.com/kubernetes/kubernetes/pull/99494/files#diff-1f1d282e9f64e36fe6a7d094b0cd8be8cb366421607c3eff23aaef6c6099193aR74

Reply:
FYI https://github.com/kubernetes/kubernetes/pull/99494 is merged now and should be in the v1.22 release.

Reply:
Thanks @enj :smile: